add an ability to restart OOP

heejaechang opened the pull request:
features that load user extension dlls to perform its functionality want a way to recycle OOP so that it can re-load those dlls without requiring VS to be restarted.

added an ability for features to request new RemoteHost. but this is not intrusive to ones already running by letting those one to finish naturally rather than forcing them to use new host.

heejaechang commented:
retest windows_debug_vs-integration_prtest please

heejaechang commented:
@mavasani @srivatsn @panopticoncentral @rynowak can you take a look?

heejaechang commented:
ping?

CyrusNajmabadi commented:
> added an ability for features to request new RemoteHost. but this is not intrusive to ones already running by letting those one to finish naturally rather than forcing them to use new host.

How will this work with features like the persistence service?  If there are multiple servers, can they all be talking to the same persistence service?  Or will code running in one server find that the persistence service is unavailable?

--

Can you ping either Roslyn-ide or me on any architectural changes to OOP?  These changes have important ramifications for the IDE, and i want us to always be in the loop on them.

heejaechang commented:
@CyrusNajmabadi sure I can ping you when I send PR on OOP. 

feature that uses persistent service is required to work even without persist service available. so functionality wise they should keep working.

on perf side, that would be a matter of making persistent service better.

CyrusNajmabadi commented:
> feature that uses persistent service is required to work even without persist service available.

I don't see how we could possibly do that effectively.  Take, for example, FAR or NavTo.  If the persistence service is unavailable then we're going to compute indices for the entire solution *every* time you run FAR or NavTo.  That's going to be absolutely, unacceptably, bad for performance.

If you make this change, and cannot provide reasonable guarantees that the server can persist data, then IDE cannot use OOP.

heejaechang commented:
@CyrusNajmabadi persistent service already doesn't exist if two VS open same solution. so all features that uses persistent service is required to work without persistent service available. that is already requirement.

heejaechang commented:
"Take, for example, FAR or NavTo. If the persistence service is unavailable then we're going to compute indices for the entire solution every time you run FAR or NavTo. That's going to be absolutely, unacceptably, bad for performance."

I am not sure whether code has been changed or not, but before, those code was written to prepare the case where persistent service not available and cached info in memory when persistent service is not available. so cache is still there, it just uses more memory.

CyrusNajmabadi commented:
> I am not sure whether code has been changed or not, but before, those code was written to prepare the case where persistent service not available and cached info in memory

How could there be something cached in memory if the persistence service isn't there to load the computed index out of?  We'll have to recompute the entire index.  That's a massive perf hit.  And one we only want to take when the solution is truly in a 'first open' state.  After that point, we always want to load/persist the index.  And if this means that there are times when it's quite possible for the persistence service to not be available, that's a significant concern for IDE.

CyrusNajmabadi commented:
There is a vast difference between persistence not being available because the same solution is opened twice, and persistence is not available because hte solution is opened once, but because of some 'restart OOP' feature, persistence is now no longer available.

We've had this conversation before.  If you change the rate at which something happens, that is significant.  Or if you change the circumstances whereby something can happen, that is also significant.  

The IDE designs our layers based on how these things actually work.  If you end up changing how things work in VS then that can break our designs.  That's why we must be involved in these sorts of decisions.

heejaechang commented:
@CyrusNajmabadi previously, if SymbolTreeInfo or SyntaxTreeIndex can't be saved to persistent storage, it will be held in memory.

that is also required since persistent storage in current design only can save data from primary workspace, not any forked solution which also require those cache when FAR or GoTo is called through API (such as rename) which can't be saved to persistant service.

CyrusNajmabadi commented:
> @CyrusNajmabadi previously, if SymbolTreeInfo or SyntaxTreeIndex can't be saved to persistent storage, it will be held in memory.

It can only be held in memory if we've produced it (either by loading it from disk, or by going and actually computing it).  It sounds like this opens the possibility of us not being able to load it from disk.  So we will go and compute it, even though the user is just doing the normal act of opening and using VS.  

That's a problem.  You're now increasing the likelihood that a full computation of the index is needed.  This is massively costly and has serious implications for IDE and our perf goals.

heejaechang commented:
@Pilchie @srivatsn @rynowak @KevinH-MS @CyrusNajmabadi looks like Cyrus wants to discuss whether OOP should support restarting OOP to re-load user extension dlls without requiring VS to reopen.

My opinion is it should be supported. and Cyrus's concern on persisted service should be dealt with separately if it ever becomes really issue.

heejaechang commented:
ping? can we start discussion on this?

rynowak commented:
FYI - I'll be OOF on vacation starting about an hour from now so I'll be unable to follow up on this discussion for a few days. 😎 

We (Razor) will have a requirement in the future to be able to run our tooling features using the assemblies from the project's dependencies. This has many of the same challenges and limitations of how analyzers in VS work today - it's basically the same problem. If you update a NuGet package while we're running a tooling operation than then we'll need to either ask the user to restart, or have a way to 'fork' a new tooling process and abandon the current one. 

I can't really speak to the specific concerns being raised here. 

We're a long way for needing this feature right now and if we end up with the 'yellow bar' telling your to restart VS that's still leagues more advanced than what we do today.

heejaechang commented:
are we okay with check this in?

heejaechang commented:
can I get some sign off. persistent service discussion will be done separately.

heejaechang commented:
retest this please

heejaechang commented:
retest this please

heejaechang commented:
retest this please

heejaechang commented:
can I get code review for this change? @mavasani @CyrusNajmabadi @KevinH-MS @rynowak

CyrusNajmabadi commented on src/VisualStudio/Core/Def/Implementation/Remote/RemoteHostClientServiceFactory.RemoteHostClientService.cs:
```diff
@@ -239,10 +247,49 @@ private void OnConnectionChanged(object sender, bool connected)
+                    if (allowRestarting)
+                    {
+                        infoBarUIs.Add(
+                            new ErrorReportingUI("Restart OOP", ErrorReportingUI.UIKind.Button, async () =>
```
async lambda doesn't seem necessary here.

CyrusNajmabadi replied:
"Restart external process".

CyrusNajmabadi commented on src/VisualStudio/Core/Def/Implementation/Remote/RemoteHostClientServiceFactory.RemoteHostClientService.cs:
```diff
@@ -239,10 +247,49 @@ private void OnConnectionChanged(object sender, bool connected)
+                }
+
+
+                // log that remote host is restarted
```
multiple newlines.

CyrusNajmabadi commented on src/VisualStudio/Core/Def/Implementation/Remote/RemoteHostClientServiceFactory.RemoteHostClientService.cs:
```diff
@@ -239,10 +247,49 @@ private void OnConnectionChanged(object sender, bool connected)
                 }
+
+                // shutdown 
+                instance.Shutdown();
```
i find most of this code very confusing.  THere's an 'instance' but also an 'instanceTask'?  And we create the instance task here... only to shutdown the isntance?  I just don't get it and it's not clear from reading the code.

CyrusNajmabadi commented:
What does this mean for features like Nuget search which are stateful?  I.e. we expect to startup an OOP session that then runs continuously.  How is the VS expected to deal with the OOP side going away?

heejaechang commented:
@CyrusNajmabadi Nuget search already know how to deal with remote client being going away. and that actually can handle restarted host as well.

heejaechang commented:
@mavasani @CyrusNajmabadi @rynowak @KevinH-MS @jinujoseph @tmat @333fred @ivanbasov can I get review on this. I addressed all feedback. I would like to check this in before this (#17334) 

thank you.

NTaylorMullen commented:
Ya this would be pretty awesome from an extension developers perspective.

rynowak commented:
I don't have any concerns about the APIs here, it's just going to be a long while before we're ready to use something like this.

heejaechang commented:
ping?

heejaechang commented:
ping pong?

panopticoncentral commented:
:shipit:
